File: 백준/Gold/4386. 별자리 만들기/별자리 만들기.py
import math
from operator import itemgetter

def find(parent, i):
    if parent[i] == i:
        return i
    return find(parent, parent[i])

def union(parent, rank, x, y):
    xroot = find(parent, x)
    yroot = find(parent, y)
    if rank[xroot] < rank[yroot]:
        parent[xroot] = yroot
    elif rank[xroot] > rank[yroot]:
        parent[yroot] = xroot
    else :
        parent[yroot] = xroot
        rank[xroot] += 1

def kruskal(graph):
    result =[]
    i, e = 0, 0
    graph.sort(key=itemgetter(2))
    parent = []
    rank = []
    for node in range(n):
        parent.append(node)
        rank.append(0)
    while e < n -1 :
        u,v,w =  graph[i]
        i = i + 1
        x = find(parent, u)
        y = find(parent ,v)
        if x != y:
            e = e + 1     
            result.append([u,v,w])
            union(parent, rank, x, y)         
    return sum(w for u,v,w in result)

n = int(input())
stars = []
edges = []
for _ in range(n):
    x, y = map(float,input().split())
    stars.append((x,y))

for i in range(n-1):
    for j in range(i+1, n):
        edges.append((i,j,math.sqrt((stars[i][0]-stars[j][0])**2 + (stars[i][1]-stars[j][1])**2)))

print("%.2f" % kruskal(edges))

File: 백준/Gold/4386. 별자리 만들기/별자리 만들기.java
import java.util.*;

class Star {
    double x, y;

    Star(double x, double y) {
        this.x = x;
        this.y = y;
    }
}

class Edge implements Comparable<Edge> {
    int u, v;
    double w;

    Edge(int u, int v, double w) {
        this.u = u;
        this.v = v;
        this.w = w;
    }

    @Override
    public int compareTo(Edge o) {
        return Double.compare(this.w, o.w);
    }
}

class UnionFind {
    private int[] parent;
    private int[] rank;

    UnionFind(int n) {
        parent = new int[n];
        rank = new int[n];
        for (int i = 0; i < n; i++) {
            parent[i] = i;
            rank[i] = 0;
        }
    }

    int find(int x) {
        if (parent[x] != x) {
            parent[x] = find(parent[x]);
        }
        return parent[x];
    }

    void union(int x, int y) {
        int xRoot = find(x);
        int yRoot = find(y);
        if (rank[xRoot] < rank[yRoot]) {
            parent[xRoot] = yRoot;
        } else if (rank[xRoot] > rank[yRoot]) {
            parent[yRoot] = xRoot;
        } else {
            parent[yRoot] = xRoot;
            rank[xRoot]++;
        }
    }
}

public class Main {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        List<Star> stars = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            stars.add(new Star(scanner.nextDouble(), scanner.nextDouble()));
        }

        List<Edge> edges = new ArrayList<>();
        for (int i = 0; i < n - 1; i++) {
            for (int j = i + 1; j < n; j++) {
                double dist = Math.sqrt(Math.pow(stars.get(i).x - stars.get(j).x, 2) +
                        Math.pow(stars.get(i).y - stars.get(j).y, 2));
                edges.add(new Edge(i, j, dist));
            }
        }
        Collections.sort(edges);
        UnionFind uf = new UnionFind(n);
        double res = 0.0;
        for (Edge e : edges) {
            if (uf.find(e.u) != uf.find(e.v)) {
                uf.union(e.u, e.v);
                res += e.w;
            }
        }
        System.out.printf("%.2f\n", res);
    }
}
